File: src/main/java/cn/labzen/logger/logback/filter/ForcedFilter.java
package cn.labzen.logger.logback.filter;

import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.turbo.TurboFilter;
import ch.qos.logback.core.spi.FilterReply;
import cn.labzen.logger.kernel.marker.ForcedMarker;
import cn.labzen.logger.kernel.marker.MarkerWrapper;
import org.slf4j.Marker;

public class ForcedFilter extends TurboFilter {

  @Override
  public FilterReply decide(Marker marker, Logger logger, Level level, String format, Object[] params, Throwable t) {
    switch (marker) {
      case MarkerWrapper mw -> {
        if (mw.getForced() != null) {
          return FilterReply.ACCEPT;
        } else {
          return FilterReply.NEUTRAL;
        }
      }
      case ForcedMarker ignored -> {
        return FilterReply.ACCEPT;
      }
      case null, default -> {
        return FilterReply.NEUTRAL;
      }
    }
  }
}

File: src/main/kotlin/cn/labzen/logger/logback/filter/ForcedFilter.kt
package cn.labzen.logger.logback.filter

import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import ch.qos.logback.classic.turbo.TurboFilter
import ch.qos.logback.core.spi.FilterReply
import cn.labzen.logger.kernel.marker.ForcedMarker
import cn.labzen.logger.kernel.marker.LabzenMarkerWrapper
import org.slf4j.Marker

class ForcedFilter : TurboFilter() {

  override fun decide(
    marker: Marker?,
    logger: Logger?,
    level: Level?,
    format: String?,
    params: Array<out Any>?,
    t: Throwable?
  ): FilterReply {
    return marker?.let {
      when (it) {
        is LabzenMarkerWrapper -> if (it.forced != null) FilterReply.ACCEPT else FilterReply.NEUTRAL
        is ForcedMarker -> FilterReply.ACCEPT
        else -> FilterReply.NEUTRAL
      }
    } ?: FilterReply.NEUTRAL
  }
}
